arangod/RocksDBEngine/RocksDBCollection.cpp

```diff
@@ -922,25 +923,31 @@ Result RocksDBCollection::truncateWithRangeDelete(transaction::Methods& trx) {
   if (!s.ok()) {
     return rocksutils::convertStatus(s);
   }
+  // we need crash server if any failure in order to trigger recovery,
+  // as rocksdb truncate already committed here
+  [&]() noexcept {
+    // post commit sequence
+    rocksdb::SequenceNumber seq = db->GetLatestSequenceNumber() - 1;
 
-  rocksdb::SequenceNumber seq =
-      db->GetLatestSequenceNumber() - 1;  // post commit sequence
-
-  uint64_t numDocs = _meta.numberDocuments();
-  _meta.adjustNumberDocuments(seq,
-                              /*revision*/ _logicalCollection.newRevisionId(),
-                              -static_cast<int64_t>(numDocs));
+    uint64_t numDocs = _meta.numberDocuments();
+    _meta.adjustNumberDocuments(seq,
+                                /*revision*/ _logicalCollection.newRevisionId(),
+                                -static_cast<int64_t>(numDocs));
 
-  {
-    RECURSIVE_READ_LOCKER(_indexesLock, _indexesLockWriteOwner);
-    for (std::shared_ptr<Index> const& idx : _indexes) {
-      idx->afterTruncate(seq,
-                         &trx);  // clears caches / clears links (if applicable)
+    {
+      RECURSIVE_READ_LOCKER(_indexesLock, _indexesLockWriteOwner);
+      auto it = guards.begin();
```

Comment:
I just realized that there is a potential race here.

We are creating guard objects for every index under the `_indexesLock` above.
Then we give up the `_indexesLock`. 

And here we acquire the `_indexesLock` again and assume that there are as many guard objects as there are indexes.
But IMHO this assumption is not necessarily true, for example, if one adds or drops an index between the two steps.

Reply:
Agreed, it doesn't affect 3.11/devel because index snapshot used instead of read lock twice.

I think this bug already was for <= 3.10 (we called remove batch on one indexes and called afterTruncate for other).
So I think better to fix it in the separate PR.

Reply:
Fine to fix this in a separate PR.